Add logging and an admin notice for Legacy REST API usages

Konamiman opened the pull request:
### Changes proposed in this Pull Request:

In order to help users determine how the Legacy REST API is used on its site in the context of [its removal in WooCommerce 9.0](https://developer.woo.com/2023/10/03/the-legacy-rest-api-will-move-to-a-dedicated-extension-in-woocommerce-9-0/), this pull request adds a helper tool that introduces two mechanisms:

1. Logging (in log files with a dedicated name) usages of the Legacy REST API, including API version, request route, user agent (when available) and date.
2. Displaying an admin notice with summary information about the requests received.

![image](https://github.com/woocommerce/woocommerce/assets/937723/b5071888-6376-4a93-8bc4-1ffa635e4231)

Internally, a transient named `wc_legacy_rest_api_usages` (with a lifetime of two weeks) is kept with the date of the first access, the total accesses count, and the different user agents registered.

The notice will not show, and logging won't happen, if any of these is true:

* The separate Legacy REST API extension is installed and active.
* The Legacy REST API is disabled.
* The transient with usages data is not available.

Additionally, it's possible to deactivate the writing of the log entry by having the `woocommerce_log_legacy_rest_api_usages` filter return `false`.

### How to test the changes in this Pull Request:

To test the changes in this pull request you need to perform requests to at least one endpoint of the Legacy REST API using Postman, `curl` or your favorite HTTP request tool. For example: `http://localhost/wc-api/v3/orders`.

1. Install [the Legacy REST API plugin](https://github.com/woocommerce/woocommerce-legacy-rest-api) but don't activate it. If you have it already installed and active, deactivate it.

2. Enable the Legacy REST API (in WooCommerce - Settings - Advanced - Legacy API) if it's not active.

2. Perform two Legacy REST API requests specifying different user agents, and then one more without user agent header.

3. Go to the logs page, verify that the notice has appeared and displays correct information (it should say that it has recognized threee accesses from two different known user agents) and that the log file has been created with the appropriate entries (the log file name starts with `legacy_rest_api_usages`).

4. Check the transient file contents with `wp transient get 'wc_legacy_rest_api_usages'`, then delete it with `wp transient delete wc_legacy_rest_api_usages` to simulate the transient expiration.

5. Reload any admin page. See how the notice has disappeared.

6. Perform yet another Legacy REST API request and reload any admin page, see how the notice is there again but its data has been reset (first date is current date and access count is 1).

7. Verify that disabling the Legacy REST API in settings makes the notice disappear, and enabling it again makes the notice appear again.

8. Activate the Legacy REST API plugin and reload the admin page, the notice has now disappeared. Verify also that Legacy REST API usages are no longer logged.

9. Deactivate the Legacy REST API plugin and reload the admin page, the notice will appear again and logging of new Legacy REST API usages will resume.

10. Dismiss the notice by clicking its "dismiss" link, it will no longer appear regardless of new Legacy REST API usages or extension activation or deactivation. If you want to undo the dismissal run this: `wp user meta delete <your user id> "dismissed_legacy_api_usages_detected_notice"`

11. Verify that Legacy REST API usages stop being logged (even without the extension active) when you add the following code: `add_filter('woocommerce_log_legacy_rest_api_usages', '__return_false');`


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Konamiman committed:
Add logging and admin noticing for rest api usages

Two new settings are added (UI in the Legacy API settings page):

- woocommerce_legacy_api_log_enabled
- woocommerce_legacy_api_usage_notice_enabled

When any of the two are enabled, legacy API usages are stored
in two options, 'wc_legacy_rest_usages' and 'wc_legacy_rest_last_usage'.

'wc_legacy_rest_usages' is a dictionary keyed by user agent,
each entry is in turn a dictionary keyed by request route,
items are arrays containing first and last usage dates as well
as total usages count (API version is logged too but for simplicity
it's not used to key the data).

'wc_legacy_rest_last_usage' contains the entry for the last usage,
regardless of user agent and route. It's used to display the notice.

When 'woocommerce_legacy_api_usage_notice_enabled' is enabled,
and the 'wc_legacy_rest_last_usage' option exists, an admin notice
displaying its contents is shown.

Konamiman committed:
Add changelog file

Konamiman commented on plugins/woocommerce/includes/legacy/class-wc-legacy-api.php:
```diff
@@ -47,6 +51,8 @@ class WC_Legacy_API {
 	 */
 	public function init() {
 		add_action( 'parse_request', array( $this, 'handle_rest_api_requests' ), 0 );
+		self:$this->mark_method_as_accessible( 'display_legacy_wc_api_usage_notice' );
```
I don't know why, but in this particular file this is needed, otherwise when the action is triggered a "Call to private method" fatal error is thrown 🤷🏻‍♀️

Konamiman committed:
Linting fixes

Konamiman committed:
Fix unit test

jorgeatorres commented on plugins/woocommerce/includes/legacy/class-wc-legacy-api.php:
```diff
@@ -62,6 +68,58 @@ public function add_query_vars( $vars ) {
+	 * @param string $route The Legacy REST API route requested.
+	 * @param string|null $user_agent The content of the user agent HTTP header in the request, null if not available.
+	 */
+	private function maybe_log_rest_api_request(string $route, ?string $user_agent) {
```
Formatting in this function is all messed up, with wrong spacing, `=` and `=>` alignment, etc. Not sure why CI isn't picking this up, though 🤷

Konamiman replied:
The beautifier/linter acted really weird locally too, and then there's the thing with the `AccessiblePrivateMethods` trait. I don't know what's so special about this file.

jorgeatorres reviewed (changes requested):
Hi @Konamiman!

Thanks for submitting this one! 💯. There are some formatting issues that need to be addressed, but on an initial look, things seem to be working as expected.

I do have some thoughts/questions/ideas on the many steps to enable some monitoring that this adds:

- I think ideally we would have _some_ monitoring on by default so we start capturing things as soon as possible without having to enable anything. Even if we now have the ability to point users to these new settings, we've already lost an opportunity to detect some usage after WC has been updated.
- Even if people enable logging, there's still no way to quickly confirm that the legacy REST API is being used, so I think we definitely need to surface that on this settings screen but ideally also in the SSR.
- Writing to the log file might be considered "heavy" in the context of a REST API request so I understand why we would want to make it a setting, but IMHO there's something we could do regardless of logging: we could use a transient or option to at least be able to quickly tell users whether the legacy REST API is even being used (and least up to that point) and ask them to enable logging/monitoring if they want more details.
- I get that we don't want to clutter log files, but maybe that's not a bad thing as that reflects the real usage on the site. We're also not enabling this by default, so it's something merchants can disable if cluttering becomes a problem. If I enable logging of something and only see a "random" selection of what I expected to see, that's weird too.

I'm not a usability expert but I think the ideal situation would be something like this:

- We record # of uses of the legacy REST API and maybe relevant dates (first/last use) without necessarily storing more info. This is enabled by default and not controlled by any setting.
- With the above we at least detect usage and are able to provide useful info as soon as we have it. For example, on the settings screen we could use this to display how many times the legacy REST API has been used since X day and also include this info in the SSR.
- I'd advocate for logging-on-by-default (maybe controlled by a filter instead of a setting) but this proves too problematic, I'd still think consolidating both settings into just one: "Monitor usage of legacy REST API endpoints" would be clearer and tidier.
- I'm not sure if the admin notice should appear everywhere (though we could make it dismissible) but in any case, I don't think it should be optional. 

What are your thoughts on the above? These are clearly not technical issues with your code, but more usability-related. If you prefer we bring this to the team and brainstorm together, let me know.

Konamiman commented:
@jorgeatorres This pull request is basically created in response to the "I have the legacy REST API enabled and I don't know why" comments in [the API removal announcement post](https://developer.woo.com/2023/10/03/the-legacy-rest-api-will-move-to-a-dedicated-extension-in-woocommerce-9-0/), so the idea was to announce the availability of the logging mechanism as a comment in that post (and possibly in an additional dedicated post too). That's why it's disabled by default: the idea is that _whoever needs it_ will enable it, and everyone else won't be "disturbed" (otherwise I  bet there would be complaints in the opposite direction: "Yes I know my legacy REST API is in use, why am I being nagged about it?")

>  I'd still think consolidating both settings into just one: "Monitor usage of legacy REST API endpoints" would be clearer and tidier

I made two settings because I considered that there will be people who wants to be able to check logs, but don't want an admin notice to be always visible in the admin area. Or the other way around: they want to see a notice just to check if the API is being used at all, bu they aren't interested in logs.

> we could use a transient or option to at least be able to quickly tell users whether the legacy REST API is even being used

That option already exists. What we could do is add a "enable logging if you want more details" link in the notice, provided that logging isn't already enabled.

> I get that we don't want to clutter log files, but maybe that's not a bad thing as that reflects the real usage on the site

We are already storing the total usage count for each combination of user agent and route, maybe it's just a matter of making this more visible somehow?

> f you prefer we bring this to the team and brainstorm together, let me know.

That's a good idea, yes.

barryhughes commented:
I tend to agree with the direction of @jorgeatorres's thoughts.

The amount of feedback we received [over here](https://developer.woo.com/2023/10/03/the-legacy-rest-api-will-move-to-a-dedicated-extension-in-woocommerce-9-0/) took me by surprise, but I'm glad people took the time to share their concerns because now we know lots of less-technical (or simply busy) users need additional support as we press ahead with the Legacy REST API removal.

> So the idea was to announce the availability of the logging mechanism as a comment in that post (and possibly in an additional dedicated post too).

Definitely—we should continue to communicate loudly. Even so, the reality is not everyone reads the developer blog (which is part of the reason I was surprised at the volume of feedback we received, largely from non-developers). 

> Yes I know my legacy REST API is in use, why am I being nagged about it?

We certainly could receive this sort of feedback ... but so long as there is a means of dismissing the notice, I think that's fine. 

For those who don't read the developer blog and don't have a reason to visit **WooCommerce ▸ Settings ▸ Advanced ▸ Legacy API**, though, the risk is they will be caught unawares. All that to say, I strongly feel the notice should be enabled by default.

jorgeatorres commented:
Based on the discussion so far, and thinking we want to provide data as soon as possible so that customers can take action, I'd like to propose we take a slightly different approach:

- No new settings are added.
- We start detecting and logging requests to the legacy REST API as soon as possible. I don't think it's worth trying to "reduce" the number requests in the logs, not only because it gives a false idea of usage but also because it means we have to keep track of things in a (possibly huge) option which doesn't look very scalable either.
- The transient/option doesn't need to contain information about all different requests. I propose we only store the following info: date of first request we've detected, the total number of requests and the date or details of the last one.
- On the settings screen we use the info above to display something like "N requests were detected since X. Last one was on Z. Check logs (link) for details."
- The admin notice is not controlled by a setting and is on by default. As soon as we detect legacy requests, we display the admin notice, which should include links to the docs, logs and details about the total number of requests and date of the last one.
   - Maybe we define some threshold here. For example, if no legacy request has been made in X days, we no longer display the admin notice, but info is still available on the settings screen.
   - We make the notice (partially) dismissible so that admins can hide it. We could provide buttons like "Ignore for/remind me in X days" or even one to fully disable logging and dismiss.
- If the legacy plugin is installed, no logging takes place and no admin notice is displayed.

@Konamiman @barryhughes: what do you think of this possible implementation?

Konamiman commented:
> I don't think it's worth trying to "reduce" the number requests in the logs

We need _some_ kind of rate limiting, or the logs can easily get flooded.

> The transient/option doesn't need to contain information about all different requests. I propose we only store the following info: date of first request we've detected, the total number of requests and the date or details of the last one.

If we don't store information about the endpoints hit, at least we should keep a list of user agents to help with the "why do I have the legacy API active" cases.

> The admin notice is not controlled by a setting and is on by default.

I guarantee you that part of the Woo users will complain about the "noise" and will request the notice to be removed. I'd leave the setting in place, but agreed, it can be on by default.

Also I still see value in having separate settings for the logging and the notice. Once the admin is aware that the legacy API is in use, they can disable the notice, but still keep the logging on to collect user agents information over time.

jorgeatorres commented:
> We need some kind of rate limiting, or the logs can easily get flooded.

I think you might be right and I'm just underestimating the number of requests that might hit those deprecated endpoints.

How about we make this a 2-week transient instead? We can define the time interval, but that's an example.
That way we can still rate limit logging, detect usage (to be displayed on the settings screen or to trigger the notice), but at least if nothing has happened in a given period (say because the merchant updated or disabled those integrations) at least the transient will fade away from the site, as opposed to a normal option.

Also, not sure why need to store "the last" request. Couldn't we just use the other transient? I don't think the notice _needs_ to be particularly about the last usage. As long as we have the other transient we know there's been some usage in X days and we can display a notice (even if a bit more generic). The details can be found in the logs.

> I guarantee you that part of the Woo users will complain about the "noise" and will request the notice to be removed. I'd leave the setting in place, but agreed, it can be on by default.

I agree, that's why I proposed we make it dismissible. As in, it contains a button like "Dismiss" or "Ignore for X days". It achieves the same thing, but we don't have to add yet another setting just for that.

> Also I still see value in having separate settings for the logging and the notice. Once the admin is aware that the legacy API is in use, they can disable the notice, but still keep the logging on to collect user agents information over time.

I might be missing something but if we make reasonable use of the transient and logging, we could be logging useful information (with no flooding) without the need to expose new settings. Allowing users to dismiss the notice would be pretty much the same as "disabling the notice" but with no user-facing setting involved.

Those new settings would be useful if we want to let people disable everything, but I don't think we should allow that. The whole point is precisely to let them know about this.

If absolutely necessary we could offer a filter to turn everything off, but my preference would be that installing the legacy plugin has that same effect. As long as merchants haven't done anything, we should be logging and possibly displaying the notice.

What do you think?

barryhughes commented:
> We need some kind of rate limiting, or the logs can easily get flooded.

&hellip;

> I think you might be right and I'm just underestimating the number of requests that might hit those deprecated endpoints.

On a number of occasions we've seen 'real' floods that stem from a warning or error being triggered in some code path leading to many-warnings-per-request ... and the logging system has generally held up admirably (albeit, it becomes hard to find information in all the noise). 

Whereas, here, there would be a hard maximum of one log entry per request (and I assume only a subset of all requests would be legacy REST API requests). My instinct then is that logging everything will not be problematic, and we don't need to engineer around it.

That said, there's clearly a range of different feelings on this, and we also don't have perfect foresight, some compromise like you're suggesting sounds pretty reasonable, fwiw.

Konamiman committed:
Simplify the approach to logging/noticing:

- Remove settings
- Use a transient instead of an option for temporary data
- Store temporary data by user agent but not by route
- Make the admin notice dismissable
- Don't log now show the notice if the legacy REST API extension
  is installed and active
- Add a filter to explicitly disable the logging

Konamiman committed:
Merge branch 'trunk' into add-legacy-rest-api-usage-logging

jorgeatorres commented on plugins/woocommerce/includes/admin/class-wc-admin-notices.php:
```diff
@@ -280,6 +280,17 @@ private static function hide_notice( $name ) {
+	 * @param string   $name The name of the admin notice to check.
+	 * @param int|null $user_id User id, or null for the current user.
+	 * @return bool True if the user has dismissed the notice.
+	 */
```
Could we add a `@since` tag to this dobclock?

jorgeatorres commented on plugins/woocommerce/includes/legacy/class-wc-legacy-api.php:
```diff
@@ -47,6 +51,8 @@ class WC_Legacy_API {
 	 */
 	public function init() {
 		add_action( 'parse_request', array( $this, 'handle_rest_api_requests' ), 0 );
+		self:$this->mark_method_as_accessible( 'maybe_display_legacy_wc_api_usage_notice' );
```
Shouldn't this just be `$this->mark_method_as_accessible()`?

jorgeatorres commented on plugins/woocommerce/includes/legacy/class-wc-legacy-api.php:
```diff
@@ -47,6 +51,8 @@ class WC_Legacy_API {
 	public function init() {
 		add_action( 'parse_request', array( $this, 'handle_rest_api_requests' ), 0 );
+		self:$this->mark_method_as_accessible( 'maybe_display_legacy_wc_api_usage_notice' );
+		self::add_action( 'admin_notices', array( $this, 'maybe_display_legacy_wc_api_usage_notice' ), 0);
```
```suggestion
		self::add_action( 'admin_notices', array( $this, 'maybe_display_legacy_wc_api_usage_notice' ), 0 );
```

jorgeatorres commented on plugins/woocommerce/includes/legacy/class-wc-legacy-api.php:
```diff
@@ -90,30 +146,72 @@ public function handle_rest_api_requests() {
+			// Fire off the request.
+			$this->server->serve_request();
+		}
+	}
```
I believe we're missing an `exit` before the function ends, to fully emulate the behavior currently in trunk:  https://github.com/woocommerce/woocommerce/blob/06e5b43d5226f8814debbff5eb54b898117f672f/plugins/woocommerce/includes/legacy/class-wc-legacy-api.php#L116

jorgeatorres commented on plugins/woocommerce/includes/legacy/class-wc-legacy-api.php:
```diff
@@ -90,30 +146,72 @@ public function handle_rest_api_requests() {
+		wc_maybe_define_constant( 'WC_API_REQUEST_VERSION', absint( $wp->query_vars['wc-api-version'] ) );
+
+		$route =  $wp->query_vars['wc-api-route'];
+		$this->maybe_log_rest_api_request($route, $_SERVER['HTTP_USER_AGENT'] ?? null);
```
```suggestion
		$this->maybe_log_rest_api_request($route, $_SERVER['HTTP_USER_AGENT'] ?? null );
```

jorgeatorres commented on plugins/woocommerce/includes/legacy/class-wc-legacy-api.php:
```diff
@@ -90,30 +146,72 @@ public function handle_rest_api_requests() {
+					),
+					sprintf(
+					// translators: Placeholders are URLs.
+						wpautop( wp_kses_data( __( '<p>The WooCommerce Legacy REST API has been accessed <b>%1$d</b> times since <b>%2$s</b>. There are <b>%3$d</b> known user agents registered. There are more details in <b><a target="_blank" href="%4$s">the WooCommerce log files</a></b> (file names start with <code>legacy_rest_api_usages</code>).', 'woocommerce' ) ) ),
```
Given there are several numbers involved, I understand it's difficult to pass this through `_n()` properly, but we could use text like "time(s)" and "user agent(s)" to sort of indicate that the number might or might not be plural.

jorgeatorres commented on plugins/woocommerce/includes/legacy/class-wc-legacy-api.php:
```diff
@@ -90,30 +146,72 @@ public function handle_rest_api_requests() {
+						esc_html__( 'WooCommerce Legacy REST API access detected', 'woocommerce' )
+					),
+					sprintf(
+					// translators: Placeholders are URLs.
```
This is not correct. Not all placeholders are URLs.

jorgeatorres commented on plugins/woocommerce/includes/legacy/class-wc-legacy-api.php:
```diff
@@ -90,30 +146,72 @@ public function handle_rest_api_requests() {
+	private function maybe_display_legacy_wc_api_usage_notice(): void {
+		if ( is_plugin_active( 'woocommerce-legacy-rest-api/woocommerce-legacy-rest-api.php' ) ) {
+			if ( WC_Admin_Notices::has_notice( 'legacy_api_usages_detected' ) ) {
+				WC_Admin_Notices::remove_notice( 'legacy_api_usages_detected', true );
```
I think we should also remove the notice if the transient doesn't exist. Let's say you actually fix your integrations and no new legacy requests are happening.
2 weeks after that WP would delete the transient itself, yet this warning will never go away unless you explicitly click "Dismiss". I believe if the transient doesn't exist, it'd be good to clear the notice too.

Similarly, we shouldn't probably show this if the Legacy REST API is actually disabled.

Konamiman replied:
> Yet this warning will never go away unless you explicitly click "Dismiss"

Yes, I implemented it that way on purpose. Removing the Legacy REST API on a system where it is in use is a big deal, and I prefer to be "too noisy" and force the user to explicitly acknowlegde the information by either dismissing the notice or installing the extension.

> Similarly, we shouldn't probably show this if the Legacy REST API is actually disabled

Same reasoning, the legacy REST API might be disabled but the notice appearing means that it has been enabled (and in use) in the past.

jorgeatorres replied:
I understand that but I think this can be confusing.

If nothing has happened in 2 weeks, why show the message? Yeah, it was enabled in the past but no request has come through in 2 weeks. If that happens, the notice will re-appear, so we're not hiding any information.

Also, if a merchant explicitly disables the legacy REST API, then showing the notice could give them the idea that it wasn't truly disabled and requests are still coming. That is, unless they are keeping track of the date they disabled it and compare it with the one in the notice.

I don't think this is necessarily a blocker, but I do think it's a weird UX.

Konamiman replied:
Ok, I have modified the code so that if either the transient exists, the legacy API is enabled or the separate extension is active, the notice won't appear.

jorgeatorres commented on plugins/woocommerce/includes/legacy/class-wc-legacy-api.php:
```diff
@@ -62,6 +68,56 @@ public function add_query_vars( $vars ) {
+		 * @since 8.5.0
+		 */
+		if ( apply_filters( 'woocommerce_log_legacy_rest_api_usages', true ) ) {
+			wc_get_logger()->info( 'LEGACY REST API USAGE DETECTED (version ' . WC_API_REQUEST_VERSION . "): $route (User agent: $user_agent)", array( 'source' => 'legacy_rest_api_usages' ) );
```
There's obviously nothing wrong with this format, but maybe we could one that would be "easier" to parse or tabulate?
For example instead of

```
2023-12-19T13:34:21+00:00 INFO LEGACY REST API USAGE DETECTED (version 3): /orders (User agent: RapidAPI/4.2.0 (Macintosh; OS X/14.2.0) GCDHTTPRequest)
```

We could log something like

```
2023-12-19T13:34:21+00:00 INFO Version: 3, Route: /orders, User agent: RapidAPI/4.2.0 (Macintosh; OS X/14.2.0) GCDHTTPRequest
```

Not that it'd make a _ton_ of difference but it'd be slightly easier to parse (in case merchants want to export to Excel or something). Also since this is already inside the legacy usages log, maybe we can drop the "LEGACY REST API USAGE DETECTED" bit as it could be considered redundant.

jorgeatorres reviewed (changes requested):
Hey @Konamiman!

Good job on this one 💯. It tests well and I really like how it's working and almost auto-magically thanks to the simplification in settings/etc. A few things:

- There are a few PHPCS issues. I noticed a few and indicated those in the review, but there might be others. Not sure if the CI check is not picking those up, but maybe you can run it manually in the affected files.
- I left some suggestions/comments here and there. Please let me know what you think.
- Would it be too hard to also surface that there are legacy REST API usages in WC > Settings > Advanced > Legacy API? I know we have the notice, but if it's dismissed, there's no way to quickly check this unless you go looking for transients manually. Maybe a simple warning similar to the one we have about the legacy REST API removal.
  For clarity, I'm talking about this screen:
  <img width="1085" alt="Screenshot 2023-12-19 at 13 42 35" src="https://github.com/woocommerce/woocommerce/assets/184724/503ec87e-b79a-4852-8d3f-c036fe698d8b">

Other than that, I think we're almost done. Let me know what you think.

Konamiman commented:
> Would it be too hard to also surface that there are legacy REST API usages in WC > Settings > Advanced > Legacy API? I know we have the notice, but if it's dismissed, there's no way to quickly check this unless you go looking for transients manually.

Rather than that I think we could simply extend the already existing message with "you can check Legacy REST API usages in the logs (link and log file name)".

Konamiman committed:
Small fixes, including a missing "exit" after request processing

Konamiman committed:
Apply suggestions from code review

Co-authored-by: Jorge A. Torres <jorge.torres@automattic.com>

jorgeatorres commented:
> Rather than that I think we could simply extend the already existing message with "you can check Legacy REST API usages in the logs (link and log file name)".

This makes sense to me 👍

Konamiman committed:
Fix linting issues

Konamiman commented:
> There are a few PHPCS issues.

There's something really weird going on with the `class-wc-legacy-api.php` file: the only way to have it scanned by the linter locally is moving it outside the `includes/legacy` directory, then scanning it at its new location, then moving it back. I did and performed an automatic fix, there are more issues but I don't know if it's worth addressing these in this particular pull request.

Konamiman committed:
Update the warning text under the "Enable legacy REST API" setting

Konamiman committed:
Change the rules to display the notice.

Now it won't appear if the transient isn't available
or if the Legacy REST API is disabled (or if the Legacy REST API
extension is active, as before); but if the user hasn't
explicitly dismissed the notice it will appear again if the transient
is recreated or the Legacy REST API is enabled again.

jorgeatorres reviewed (approved):
LGTM. Thanks for your hard work on this one @Konamiman! It turned out great.
